spack test run kokkos fails (possible compiler version needs pinning)

```bash
 spack test run kokkos
==> Spack test gqrsnwuo3spdzujvssfpekm4lrazskgn
==> Testing package kokkos-3.4.01-l4tyn6y
==> Error: ProcessError: Command exited with status 1:
    'cmake' '/home/vanessa/Desktop/Code/spack-vsoch/opt/spack/linux-ubuntu20.04-skylake/gcc-9.3.0/kokkos-3.4.01-l4tyn6ymtuyz5v6wu7pvuftveykpw725/.spack/test/scripts/spack_test/out' '-DEXECUTABLE_OUTPUT_PATH=/home/vanessa/Desktop/Code/spack-vsoch/opt/spack/linux-ubuntu20.04-skylake/gcc-9.3.0/kokkos-3.4.01-l4tyn6ymtuyz5v6wu7pvuftveykpw725/.spack/test/scripts/spack_test/out'

4 errors found in test log:
     7     -- Detecting CXX compiler ABI info
     8     -- Detecting CXX compiler ABI info - done
     9     -- Detecting CXX compile features
     10    -- Detecting CXX compile features - done
     11    -- Enabled Kokkos devices: SERIAL
     12    -- Configuring done
  >> 13    CMake Error at CMakeLists.txt:18 (target_link_libraries):
     14      Error evaluating generator expression:
     15    
     16        $<LINK_LANGUAGE:CXX>
     17    
     18      Expression did not evaluate to a known generator expression
     19    
     20    
  >> 21    CMake Error at CMakeLists.txt:18 (target_link_libraries):
     22      Error evaluating generator expression:
     23    
     24        $<LINK_LANGUAGE:CXX>
     25    
     26      Expression did not evaluate to a known generator expression
     27    
     28    
  >> 29    CMake Error at CMakeLists.txt:18 (target_link_libraries):
     30      Error evaluating generator expression:
     31    
     32        $<LINK_LANGUAGE:CXX>
     33    
     34      Expression did not evaluate to a known generator expression
     35    
     36    
  >> 37    CMake Error at CMakeLists.txt:18 (target_link_libraries):
     38      Error evaluating generator expression:
     39    
     40        $<LINK_LANGUAGE:CXX>
     41    
     42      Expression did not evaluate to a known generator expression
     43    

See test log for details:
  /home/vanessa/.spack/test/gqrsnwuo3spdzujvssfpekm4lrazskgn/kokkos-3.4.01-l4tyn6y-test-out.txt

==> Error: 1 test(s) in the suite failed.
```

--- a/var/spack/repos/builtin/packages/kokkos/package.py
+++ b/var/spack/repos/builtin/packages/kokkos/package.py
@@ -331,7 +331,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
 
     def build_tests(self):
         """Build test."""
-        cmake_path = join_path(self.install_test_root,
+        cmake_path = join_path(self.test_suite.current_test_cache_dir,
                                self.test_script_relative_path, 'out')
         cmake_bin = self.cmake_bin(set=False)
 
@@ -341,8 +341,6 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
 
         cmake_args = [cmake_path, '-DEXECUTABLE_OUTPUT_PATH=' + cmake_path,
                       cmake_bin]
-
-        tty.msg(type(cmake_args))
  
         if not self.run_test(cmake_bin,
                              options=cmake_args,
@@ -358,7 +356,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
     def run_tests(self):
         """Run test."""
         reason = 'Checking ability to execute.'
-        run_path = join_path(self.install_test_root,
+        run_path = join_path(self.test_suite.current_test_cache_dir,
                              self.test_script_relative_path, 'out')
         self.run_test('make',
                       options=[run_path, 'test'],
@@ -366,7 +364,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
 
     def test(self):
         # Skip if unsupported version
-        cmake_path = join_path(self.install_test_root,
+        cmake_path = join_path(self.test_suite.current_test_cache_dir,
                                self.test_script_relative_path, 'out')
         if not os.path.exists(cmake_path):
             print('Skipping smoke tests: {0} is missing'.format(cmake_path))

--- a/var/spack/repos/builtin/packages/kokkos/package.py
+++ b/var/spack/repos/builtin/packages/kokkos/package.py
@@ -4,6 +4,8 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 import os.path
 
+from llnl.util import tty
+
 from spack import *
 
 
@@ -294,6 +296,20 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
 
     test_script_relative_path = "scripts/spack_test"
 
+    # TODO: Replace this method and its 'get' use for cmake path with
+    #   join_path(self.spec['cmake'].prefix.bin, 'cmake') once stand-alone
+    #   tests can access build dependencies through self.spec['cmake'].
+    def cmake_bin(self, set=True):
+        """(Hack) Set/get cmake dependency path."""
+        filepath = join_path(self.install_test_root, 'cmake_bin_path.txt')
+        if set:
+            with open(filepath, 'w') as out_file:
+                cmake_bin = join_path(self.spec['cmake'].prefix.bin, 'cmake')
+                out_file.write('{0}\n'.format(cmake_bin))
+        elif os.path.isfile(filepath):
+            with open(filepath, 'r') as in_file:
+                return in_file.read().strip()
+
     @run_after('install')
     def setup_build_tests(self):
         # Skip if unsupported version
@@ -311,21 +327,42 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
                       "-DSPACK_PACKAGE_INSTALL_DIR:PATH={0}".format(self.prefix)]
         cmake(*cmake_args)
         self.cache_extra_test_sources(cmake_out_path)
+        self.cmake_bin(set=True)
 
     def build_tests(self):
         """Build test."""
         cmake_path = join_path(self.install_test_root,
                                self.test_script_relative_path, 'out')
-        cmake_args = [cmake_path, '-DEXECUTABLE_OUTPUT_PATH=' + cmake_path]
-        cmake(*cmake_args)
-        make()
+        cmake_bin = self.cmake_bin(set=False)
+
+        if not cmake_bin:
+            tty.msg('Skipping kokkos test: cmake_bin_path.txt not found')
+            return
+
+        cmake_args = [cmake_path, '-DEXECUTABLE_OUTPUT_PATH=' + cmake_path,
+                      cmake_bin]
+
+        tty.msg(type(cmake_args))
+ 
+        if not self.run_test(cmake_bin,
+                             options=cmake_args,
+                             purpose='Generate the Makefile'):
+            tty.msg('Skipping kokkos test: failed to generate Makefile')
+            return
+
+        if not self.run_test('make',
+                             purpose='Build test software'):
+            tty.msg('Skipping kokkos test: failed to build test')
+            return
 
     def run_tests(self):
         """Run test."""
         reason = 'Checking ability to execute.'
         run_path = join_path(self.install_test_root,
                              self.test_script_relative_path, 'out')
-        self.run_test('make', [run_path, 'test'], [], installed=False, purpose=reason)
+        self.run_test('make',
+                      options=[run_path, 'test'],
+                      purpose=reason)
 
     def test(self):
         # Skip if unsupported version

--- a/var/spack/repos/builtin/packages/kokkos/package.py
+++ b/var/spack/repos/builtin/packages/kokkos/package.py
@@ -294,7 +294,7 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
 
         return options
 
-    test_script_relative_path = "scripts/spack_test"
+    test_script_relative_path = join_path('scripts', 'spack_test')
 
     # TODO: Replace this method and its 'get' use for cmake path with
     #   join_path(self.spec['cmake'].prefix.bin, 'cmake') once stand-alone
@@ -329,45 +329,41 @@ class Kokkos(CMakePackage, CudaPackage, ROCmPackage):
         self.cache_extra_test_sources(cmake_out_path)
         self.cmake_bin(set=True)
 
-    def build_tests(self):
+    def build_tests(self, cmake_path):
         """Build test."""
-        cmake_path = join_path(self.test_suite.current_test_cache_dir,
-                               self.test_script_relative_path, 'out')
         cmake_bin = self.cmake_bin(set=False)
 
         if not cmake_bin:
             tty.msg('Skipping kokkos test: cmake_bin_path.txt not found')
             return
 
-        cmake_args = [cmake_path, '-DEXECUTABLE_OUTPUT_PATH=' + cmake_path,
-                      cmake_bin]
- 
+        cmake_args = [cmake_path, '-DEXECUTABLE_OUTPUT_PATH=' + cmake_path]
+
         if not self.run_test(cmake_bin,
                              options=cmake_args,
                              purpose='Generate the Makefile'):
-            tty.msg('Skipping kokkos test: failed to generate Makefile')
+            tty.warn('Skipping kokkos test: failed to generate Makefile')
             return
 
         if not self.run_test('make',
                              purpose='Build test software'):
-            tty.msg('Skipping kokkos test: failed to build test')
-            return
+            tty.warn('Skipping kokkos test: failed to build test')
 
-    def run_tests(self):
+    def run_tests(self, cmake_path):
         """Run test."""
-        reason = 'Checking ability to execute.'
-        run_path = join_path(self.test_suite.current_test_cache_dir,
-                             self.test_script_relative_path, 'out')
-        self.run_test('make',
-                      options=[run_path, 'test'],
-                      purpose=reason)
+        if not self.run_test('make',
+                             options=[cmake_path, 'test'],
+                             purpose='Checking ability to execute.'):
+            tty.warn('Failed to run kokkos test')
 
     def test(self):
         # Skip if unsupported version
         cmake_path = join_path(self.test_suite.current_test_cache_dir,
                                self.test_script_relative_path, 'out')
+
         if not os.path.exists(cmake_path):
-            print('Skipping smoke tests: {0} is missing'.format(cmake_path))
+            tty.warn('Skipping smoke tests: {0} is missing'.format(cmake_path))
             return
-        self.build_tests()
-        self.run_tests()
+
+        self.build_tests(cmake_path)
+        self.run_tests(cmake_path)

